
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />
<script type="text/javascript">

      var _gaq = _gaq || [];
      _gaq.push(['_setAccount', 'UA-139058009-1']);
      _gaq.push(['_trackPageview']);

      (function() {
        var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
      })();
    </script>
    <title>Commands from section ‘coverage’ &#8212; gtftk 1.5.2 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/nature.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Commands from section ‘miscellaneous’" href="miscellaneous.html" />
    <link rel="prev" title="Commands from section ‘sequence’" href="sequence.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="miscellaneous.html" title="Commands from section ‘miscellaneous’"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="sequence.html" title="Commands from section ‘sequence’"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">gtftk 1.5.2 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Commands from section ‘coverage’</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <section id="commands-from-section-coverage">
<h1>Commands from section ‘coverage’<a class="headerlink" href="#commands-from-section-coverage" title="Permalink to this headline">¶</a></h1>
<p>In this section we will require the following datasets:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_example -q -d mini_real -f &#39;*&#39;
</pre></div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_example -q  -d mini_real_noov_rnd_tx -f &#39;*&#39;
</pre></div>
</div>
<section id="coverage">
<h2>coverage<a class="headerlink" href="#coverage" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Takes a GTF as input to compute bigwig coverage in regions of interest (promoter, transcript body, intron, intron_by_tx, tts…) or a BED6 to focus on user-defined regions. If –n-highest is used the program will compute the coverage of each bigwig based on the average value of the n windows (–nb-window) with the highest coverage values.
Regions were signal can be computed (if GTF file as input) are promoter, tts, introns, intergenic regions or any feature available in the GTF file (transcript, exon, gene…).
If –matrix-out is selected, the signal for each bigwig will be provided in a dedicated column. Otherwise, signal for each bigwig is provided through a dedicated line.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The coverage example below use 8 CPUs. Please adapt.</p>
</div>
<p>We will compute coverage of promoters regions using 3 bigWig files as input.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk coverage -l H3K4me3,H3K79me2,H3K36me3 -u 5000 -d 5000 -i mini_real_noov_rnd_tx.gtf.gz -c hg38 -m transcript_id,gene_name -x ENCFF742FDS_H3K4me3_K562_sub.bw ENCFF947DVY_H3K79me2_K562_sub.bw ENCFF431HAA_H3K36me3_K562_sub.bw -k 8 &gt; coverage.bed
</pre></div>
</div>
<p>Now we can have a look at the result:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ head -n 10 coverage.bed
chrom	start	end	name	strand	H3K4me3	H3K79me2	H3K36me3
chr1	996137	1006138	ENST00000624697|ISG15	+	5.859314	4.0025	1.632737
chr1	1370156	1380157	ENST00000338370|AURKAIP1	-	25.746325	12.173583	3.20328
chr1	1914257	1924258	ENST00000378598|TMEM52	-	5.927807	2.4218580000000003	1.374963
chr1	2202732	2212733	ENST00000469733|FAAP20	-	1.0	1.0	1.0
chr1	2493646	2503647	ENST00000462379|PLCH2	+	1.227377	1.062994	1.068393
chr1	3064167	3074168	ENST00000511072|PRDM16	+	1.0	1.0	1.0
chr1	3630449	3640450	ENST00000469643|WRAP73	-	1.0	1.0	1.0
chr1	6414902	6424903	ENST00000377836|HES2	-	1.951105	1.682232	1.439256
chr1	9178749	9188750	ENST00000437157|MIR34AHG	-	1.331067	1.272973	1.436956
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk coverage -h
  Usage: gtftk coverage [-i GTF/BED] [-o TXT] -c CHROMINFO [-u UPSTREAM] [-d DOWNSTREAM] [-w nb_window] [-k nb_proc] [-f ft_type] [-l labels] [-m name_column] [-p pseudo_count] [-n n_highest] [-x] [-zn] [-a key_name] [-s {mean,sum}] [-h] [-V [verbosity]] [-D] [-C] [-K tmp_dir] [-A] [-L logger_file] [-W write_message_to_file] bw_list [bw_list ...]

  Description: 

     Takes a GTF as input to compute bigwig coverage in regions of interest (promoter, transcript body,
     intron, intron_by_tx, tts...) or a BED6 to focus on user-defined regions. If --n-highest is
     used the program will compute the coverage of each bigwig based on the average value of the n
     windows (--nb-window) with the highest coverage values.

  Notes:
     *  chrom-info may also accept &#39;mm8&#39;, &#39;mm9&#39;, &#39;mm10&#39;, &#39;hg19&#39;, &#39;hg38&#39;, &#39;rn3&#39; or &#39;rn4&#39;. In this
     case the  corresponding size of conventional chromosomes are used. To get the size of  the
     chromosome in ensembl format (whithout chr prefix), use &#39;mm8_ens&#39;, &#39;mm9_ens&#39;,  &#39;mm10_ens&#39;,
     &#39;hg19_ens&#39;, &#39;hg38_ens&#39;, &#39;rn3_ens&#39; or &#39;rn4_ens&#39;. ChrM is not used.
     *  Regions were signal can be computed (if GTF file as input): promoter/tss, tts, introns,
     intron_by_tx, intergenic regions or any feature available in the GTF file (transcript, exon,
     gene...).
     *  If --matrix-out is selected, the signal for each bigwig will be provided in a dedicated
     column. Otherwise, signal for each bigwig is provided through a dedicated line.
     *  If bed is used as input, each region should have its own name (column 4).

Arguments:
 bw_list                      A list of Bigwig file (last argument).
 -i, --inputfile              The input GTF/BED file. Only GTF file if &lt;stdin&gt; is used. (default: &lt;stdin&gt;)
 -o, --outputfile             Output file. (default: &lt;stdout&gt;)
 -c, --chrom-info             Tabulated two-columns file. Chromosomes as column 1 and sizes as column 2 (default: None)
 -u, --upstream               Extend the regions in 5&#39; by a given value (int). (default: 0)
 -d, --downstream             Extend the regions in 3&#39; by a given value (int). (default: 0)
 -w, --nb-window              Split the region into w bins (see -n). (default: 1)
 -k, --nb-proc                Use this many threads to compute coverage. (default: 1)
 -f, --ft-type                Region in which coverage is to be computed (promoter, intron, intergenic, tts or any feature defined in the column 3 of the GTF). (default: promoter)
 -l, --labels                 Bigwig labels. (default: None)
 -m, --name-column            Use this ids to compute the name (4th column in bed output). (default: transcript_id)
 -p, --pseudo-count           A pseudo-count to add in case count is equal to 0. (default: 1)
 -n, --n-highest              For each bigwig, use the n windows with higher values to compute coverage. (default: None)
 -x, --matrix-out             Matrix output format. Bigwigs as column names and features as rows. (default: False)
 -zn, --zero-to-na            Use NA not zero when region is undefined in bigwig or below window size. (default: False)
 -a, --key-name               If gtf format is requested, the name of the key. (default: cov)
 -s, --stat                   The statistics to be computed for each region. (default: mean)

Command-wise optional arguments:
 -h, --help                   Show this help message and exit.
 -V, --verbosity              Set output verbosity ([0-3]). (default: 0)
 -D, --no-date                Do not add date to output file names. (default: False)
 -C, --add-chr                Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir                Keep all temporary files into this folder. (default: None)
 -A, --keep-all               Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file            Stores the arguments passed to the command into a file. (default: None)
 -W, --write-message-to-file  Store all message into a file. (default: None)
</pre></div>
</div>
</section>
<hr class="docutils" />
<section id="mk-matrix">
<h2>mk_matrix<a class="headerlink" href="#mk-matrix" title="Permalink to this headline">¶</a></h2>
<p>Description: Gtftk implements commands that can be used to produce coverage profiles around genomic features or inside user-defined regions. A coverage matrix needs first to  be produced from a bwig using the mk_matrix command.</p>
<p><strong>Example:</strong></p>
<p>We will used the same dataset (mini_real.gtf) as produced above (see help on <em>coverage</em> command).</p>
<p>We can now create a coverage matrix around TSS/TTS or along the full transcript
(with or without 5’ and 3’ regions). Provide a BED file as <em>—inputfile</em> if you
want to use your own, user-specific, regions.
Will will create tree example datasets:</p>
<p>First we will create a coverage matrix around promoter based on a subset of randomly choose transcripts (one per gene) from the ‘mini_real’ dataset (see section on the <em>coverage</em> command to get info about the construction of the <em>mini_real_noov_rnd_tx.gtf.gz</em> dataset).</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The mk_matrix example below use 8 CPUs. Please adapt.</p>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk mk_matrix -k 8 -i mini_real_noov_rnd_tx.gtf.gz -d 5000 -u 5000 -w 200 -c hg38  -l  H3K4me3,H3K79me,H3K36me3 -y ENCFF742FDS_H3K4me3_K562_sub.bw ENCFF947DVY_H3K79me2_K562_sub.bw ENCFF431HAA_H3K36me3_K562_sub.bw -o mini_real_promoter
</pre></div>
</div>
<p>The following command computes coverage profil along the whole transcript.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk mk_matrix -k 8 --bin-around-frac 0.5 -i mini_real_noov_rnd_tx.gtf.gz -t transcript  -d 5000 -u 5000 -w 200 -c hg38  -l  H3K4me3,H3K79me,H3K36me3 -y ENCFF742FDS_H3K4me3_K562_sub.bw ENCFF947DVY_H3K79me2_K562_sub.bw ENCFF431HAA_H3K36me3_K562_sub.bw -o mini_real_tx
</pre></div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk mk_matrix -h
  Usage: gtftk mk_matrix [-i GTF/BED] -y bigwiglist [bigwiglist ...] -o GTF/TXT [-l labels] [-t {promoter,tts,transcript,user_regions,single_nuc}] [-p pseudo_count] [-u upstream] [-d downstream] -c chrom_info [-w bin_nb] [-k nb_proc] [-b bin_around_frac] [-zn] [-nst] [-h] [-V [verbosity]] [-D] [-C] [-K tmp_dir] [-A] [-L logger_file] [-W write_message_to_file]

  Description: 

     Create a matrix storing the bigwig coverage computed from binned regions.

  Notes:
     *  chrom-info may also accept &#39;mm8&#39;, &#39;mm9&#39;, &#39;mm10&#39;, &#39;hg19&#39;, &#39;hg38&#39;, &#39;rn3&#39; or &#39;rn4&#39;. In this
     case the  corresponding size of conventional chromosomes are used. To get the size of  the
     chromosome in ensembl format (whithout chr prefix), use &#39;mm8_ens&#39;, &#39;mm9_ens&#39;,  &#39;mm10_ens&#39;,
     &#39;hg19_ens&#39;, &#39;hg38_ens&#39;, &#39;rn3_ens&#39; or &#39;rn4_ens&#39;. ChrM is not used.

Arguments:
 -i, --inputfile              A GTF file or bed file. A GTF if &lt;stdin&gt;. (default: &lt;stdin&gt;)
 -y, --bigwiglist             A list of Bigwig files (last argument). (default: None)
 -o, --outputfile             Output file name (.zip extension will be added). (default: &lt;stdout&gt;)
 -l, --labels                 Bigwig labels (i.e short name version for plotting). (default: None)
 -t, --ft-type                If input is a GTF, the region to analyse. (default: promoter)
 -p, --pseudo-count           Pseudo-count to add to all values. (default: 0)
 -u, --upstream               Extend the region of interest in 5&#39; by a given value. (default: 1000)
 -d, --downstream             Extend the region of interest in 3&#39; by a given value. (default: 1000)
 -c, --chrom-info             Tabulated file (chr as column 1, sizes as column 2.) (default: None)
 -w, --bin-nb                 Split the region into w bins. (default: 100)
 -k, --nb-proc                Use this many threads to compute coverage. (default: 1)
 -b, --bin-around-frac        Fraction of bins used in 5&#39; and 3&#39; regions. (default: 0.1)
 -zn, --zero-to-na            Use NA not zero when region is undefined in bigwig. (default: False)
 -nst, --no-stranded          The bins should not be oriented relative to strand. (default: False)

Command-wise optional arguments:
 -h, --help                   Show this help message and exit.
 -V, --verbosity              Set output verbosity ([0-3]). (default: 0)
 -D, --no-date                Do not add date to output file names. (default: False)
 -C, --add-chr                Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir                Keep all temporary files into this folder. (default: None)
 -A, --keep-all               Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file            Stores the arguments passed to the command into a file. (default: None)
 -W, --write-message-to-file  Store all message into a file. (default: None)
</pre></div>
</div>
</section>
<hr class="docutils" />
<section id="profile">
<h2>profile<a class="headerlink" href="#profile" title="Permalink to this headline">¶</a></h2>
<p>Description: This command is used to create profil diagrams from a <em>mk_matrix</em> output. The two important arguments for
this command are <em>—group-by</em>, that defines the variable controling the set of colored lines and <em>—facet-var</em> that defines the variable controling the way the plot is facetted. Both <em>—group-by</em> and <em>—facet-var</em> should be set to one of <em>bwig</em>, <em>tx_classes</em> or <em>chrom</em>.</p>
<p><strong>Basic profiles</strong></p>
<p>A simple overlayed profile of all epigenetic marks around promoter. Here <em>—group-by</em> is, by default set to <em>bwig</em> and <em>—facet-var</em> is set to None. Thus a single plot with several lines corresponding to bwig coverage is obtained.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk profile -D -i mini_real_promoter.zip -o profile_prom -pf png -if  example_01.png
 |-- 14:39-WARNING-profile : --group-by not set. Choosing &#39;bwig&#39;.
</pre></div>
</div>
<a class="reference external image-reference" href="_static/example_01.png"><img alt="_images/example_01.png" src="_images/example_01.png" style="width: 75%;" /></a>
<p>Changing colors and applying color order can be done using the following syntax:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk profile -D -i mini_real_promoter.zip -c &#39;red,blue,violet&#39; -d H3K79me,H3K4me3,H3K36me3 -o profile_prom -pf png -if  example_01b.png
 |-- 14:39-WARNING-profile : --group-by not set. Choosing &#39;bwig&#39;.
</pre></div>
</div>
<a class="reference external image-reference" href="_static/example_01b.png"><img alt="_images/example_01b.png" src="_images/example_01b.png" style="width: 75%;" /></a>
<p>Transcript coverage is obtained using the <em>mini_real_tx.zip</em> matrix. This provides a simple overlayed profile of all epigenetic marks along the transcript body extended in 5’ and 3’ regions:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk profile -D -i mini_real_tx.zip -o profile_tx -pf png -if  example_02.png
 |-- 14:39-WARNING-profile : --group-by not set. Choosing &#39;bwig&#39;.
</pre></div>
</div>
<a class="reference external image-reference" href="_static/example_02.png"><img alt="_images/example_02.png" src="_images/example_02.png" style="width: 75%;" /></a>
<p><strong>Faceted profiles</strong></p>
<p>Faceted plot of epigenetic profiles. The groups (i.e colors/lines) can be set to bwig classes and the facets to transcript classes. Things can be simply done by providing an additional file containing the transcript and their associated classes.</p>
<p><strong>Example:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk profile -D -i mini_real_promoter.zip -f tx_classes -g bwig  -t tx_classes.txt -o profile_prom  -pf png -if  example_05.png -e -V 2 -fc 2
 |-- 14:39-DEBUG-profile : Using pandas version 1.2.5
 |-- 14:39-DEBUG-profile : Pandas location /Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/pandas/__init__.py
 |-- 14:39-DEBUG-profile : Using numpy version 1.21.0
 |-- 14:39-DEBUG-profile : Pandas numpy /Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/numpy/__init__.py
 |-- 14:39-DEBUG-profile : Using plotnine version 0.8.0
 |-- 14:39-DEBUG-profile : Pandas plotnine /Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/plotnine/__init__.py
 |-- 14:39-DEBUG-profile : Creating directory : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_2df2e72e
 |-- 14:39-DEBUG-profile : Uncompressing : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_2df2e72e
 |-- 14:39-DEBUG-profile : Reading : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_2df2e72e/mini_real_promoter
 |-- 14:39-INFO-profile : Getting configuration info from input file.
 |-- 14:39-INFO-profile : Reading transcript file.
 |-- 14:39-INFO-profile : Deleting duplicates in transcript-file.
 |-- 14:39-INFO-profile : Checking how many transcripts where found in the transcript list.
 |-- 14:39-INFO-profile : Keeping 804 transcript out of 833 in input transcript list.
 |-- 14:39-DEBUG-profile : Color order : [&#39;H3K79me&#39;, &#39;H3K4me3&#39;, &#39;H3K36me3&#39;]
 |-- 14:39-DEBUG-profile : Profile color : [&#39;#000000&#39;, &#39;#00bb00&#39;, &#39;#cccccc&#39;]
 |-- 14:39-INFO-profile : Searching coverage columns.
 |-- 14:39-INFO-profile : Melting.
 |-- 14:39-INFO-profile : Ceiling
 |-- 14:39-INFO-profile : Computing column ordering.
 |-- 14:39-INFO-profile : Preparing diagram
 |-- 14:39-INFO-profile : Theming and ordering. Please be patient...
 |-- 14:39-INFO-profile : Preparing x axis
 |-- 14:39-INFO-profile : facet_col 2
 |-- 14:39-INFO-profile : Page width set to 6
 |-- 14:39-INFO-profile : Page height set to 5.0
 |-- 14:39-INFO-profile : Saving diagram to file : example_05.png
 |-- 14:39-INFO-profile : Be patient. This may be long for large datasets.
 |-- 14:39-DEBUG-profile : Deleting temp file : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_2df2e72e
</pre></div>
</div>
<a class="reference external image-reference" href="_static/example_05.png"><img alt="_images/example_05.png" src="_images/example_05.png" style="width: 75%;" /></a>
<p>Alternatively, the groups can be set to chromosomes or transcript classes:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk profile -D -i mini_real_promoter.zip -g tx_classes -f bwig  -t tx_classes.txt -o profile_prom  -pf png -if  example_06.png -V 2 -nm ranging
 |-- 14:39-DEBUG-profile : Using pandas version 1.2.5
 |-- 14:39-DEBUG-profile : Pandas location /Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/pandas/__init__.py
 |-- 14:39-DEBUG-profile : Using numpy version 1.21.0
 |-- 14:39-DEBUG-profile : Pandas numpy /Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/numpy/__init__.py
 |-- 14:39-DEBUG-profile : Using plotnine version 0.8.0
 |-- 14:39-DEBUG-profile : Pandas plotnine /Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/plotnine/__init__.py
 |-- 14:39-DEBUG-profile : Creating directory : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_gwqhgdui
 |-- 14:39-DEBUG-profile : Uncompressing : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_gwqhgdui
 |-- 14:39-DEBUG-profile : Reading : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_gwqhgdui/mini_real_promoter
 |-- 14:39-INFO-profile : Getting configuration info from input file.
 |-- 14:39-INFO-profile : Reading transcript file.
 |-- 14:39-INFO-profile : Deleting duplicates in transcript-file.
 |-- 14:39-INFO-profile : Checking how many transcripts where found in the transcript list.
 |-- 14:39-INFO-profile : Keeping 804 transcript out of 833 in input transcript list.
 |-- 14:39-DEBUG-profile : Color order : [&#39;protein_coding&#39;, &#39;antisense&#39;, &#39;lincRNA&#39;]
 |-- 14:39-DEBUG-profile : Profile color : [&#39;#000000&#39;, &#39;#00bb00&#39;, &#39;#cccccc&#39;]
 |-- 14:39-INFO-profile : Searching coverage columns.
 |-- 14:39-INFO-profile : Melting.
 |-- 14:39-INFO-profile : Ceiling
 |-- 14:39-INFO-profile : Normalizing (ranging)
 |-- 14:39-INFO-profile : Computing column ordering.
 |-- 14:39-INFO-profile : Preparing diagram
 |-- 14:39-INFO-profile : Theming and ordering. Please be patient...
 |-- 14:39-INFO-profile : Preparing x axis
 |-- 14:39-INFO-profile : facet_col 3
 |-- 14:39-INFO-profile : Page width set to 9
 |-- 14:39-INFO-profile : Page height set to 2.0
 |-- 14:39-INFO-profile : Saving diagram to file : example_06.png
 |-- 14:39-INFO-profile : Be patient. This may be long for large datasets.
 |-- 14:39-DEBUG-profile : Deleting temp file : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_gwqhgdui
</pre></div>
</div>
<a class="reference external image-reference" href="_static/example_06.png"><img alt="_images/example_06.png" src="_images/example_06.png" style="width: 75%;" /></a>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk profile -D -i mini_real_promoter.zip -g chrom -f bwig  -t tx_classes.txt -o profile_prom  -pf png -if  example_06b.png -V 2 -nm ranging
 |-- 14:39-DEBUG-profile : Using pandas version 1.2.5
 |-- 14:39-DEBUG-profile : Pandas location /Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/pandas/__init__.py
 |-- 14:39-DEBUG-profile : Using numpy version 1.21.0
 |-- 14:39-DEBUG-profile : Pandas numpy /Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/numpy/__init__.py
 |-- 14:39-DEBUG-profile : Using plotnine version 0.8.0
 |-- 14:39-DEBUG-profile : Pandas plotnine /Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/plotnine/__init__.py
 |-- 14:39-DEBUG-profile : Creating directory : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_sxoccetc
 |-- 14:39-DEBUG-profile : Uncompressing : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_sxoccetc
 |-- 14:39-DEBUG-profile : Reading : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_sxoccetc/mini_real_promoter
 |-- 14:39-INFO-profile : Getting configuration info from input file.
 |-- 14:39-DEBUG-profile : Color order : [&#39;chrX&#39;, &#39;chr3&#39;, &#39;chr21&#39;, &#39;chr9&#39;, &#39;chr8&#39;, &#39;chr5&#39;, &#39;chr10&#39;, &#39;chr19&#39;, &#39;chr2&#39;, &#39;chr1&#39;, &#39;chr18&#39;, &#39;chr14&#39;, &#39;chr22&#39;, &#39;chr11&#39;, &#39;chr16&#39;, &#39;chr17&#39;, &#39;chr13&#39;, &#39;chr20&#39;, &#39;chr6&#39;, &#39;chr4&#39;, &#39;chr15&#39;, &#39;chr12&#39;, &#39;chr7&#39;]
 |-- 14:39-DEBUG-profile : Profile color : [&#39;#000000&#39;, &#39;#6c007c&#39;, &#39;#850096&#39;, &#39;#2500a5&#39;, &#39;#0000ca&#39;, &#39;#0041dd&#39;, &#39;#0086dd&#39;, &#39;#009fca&#39;, &#39;#00aaa1&#39;, &#39;#00a76f&#39;, &#39;#009c00&#39;, &#39;#00bb00&#39;, &#39;#00da00&#39;, &#39;#00f900&#39;, &#39;#88ff00&#39;, &#39;#dbf400&#39;, &#39;#f7db00&#39;, &#39;#ffb500&#39;, &#39;#ff6100&#39;, &#39;#f60000&#39;, &#39;#da0000&#39;, &#39;#cc1313&#39;, &#39;#cccccc&#39;]
 |-- 14:39-INFO-profile : Searching coverage columns.
 |-- 14:39-INFO-profile : Melting.
 |-- 14:39-INFO-profile : Ceiling
 |-- 14:39-INFO-profile : Normalizing (ranging)
 |-- 14:39-INFO-profile : Computing column ordering.
 |-- 14:39-INFO-profile : Preparing diagram
 |-- 14:39-INFO-profile : Theming and ordering. Please be patient...
 |-- 14:39-INFO-profile : Preparing x axis
 |-- 14:39-INFO-profile : facet_col 3
 |-- 14:39-INFO-profile : Page width set to 9
 |-- 14:39-INFO-profile : Page height set to 2.0
 |-- 14:39-INFO-profile : Saving diagram to file : example_06b.png
 |-- 14:39-INFO-profile : Be patient. This may be long for large datasets.
 |-- 14:39-DEBUG-profile : Deleting temp file : /var/folders/zy/wl3dj2_n76zfc8sdvny1q06c0000gn/T/profile_matrix__pygtftk_sxoccetc
</pre></div>
</div>
<a class="reference external image-reference" href="_static/example_06b.png"><img alt="_images/example_06b.png" src="_images/example_06b.png" style="width: 75%;" /></a>
<p>Note that facets may also be associated to epigenetic marks. In this case each the –group-by can be set to <em>tx_classes</em> or <em>chrom</em>.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk profile -D -i mini_real_tx.zip -g tx_classes -t tx_classes.txt -f bwig  -o profile_tx -pf png -if  example_07.png -w -nm ranging
/Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/pygtftk-1.5.2-py3.8-macosx-10.9-x86_64.egg/pygtftk/plugins/profile.py:890: FutureWarning: Index.ravel returning ndarray is deprecated; in a future version this will return a view on self.
 |-- 14:39-WARNING-profile : PlotnineError error, can not show group number: Aesthetics {&#39;ha&#39;} specified two times.
</pre></div>
</div>
<a class="reference external image-reference" href="_static/example_07.png"><img alt="_images/example_07.png" src="_images/example_07.png" style="width: 75%;" /></a>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk profile -D -i mini_real_tx.zip -g chrom -f bwig  -o profile_tx -pf png -if  example_08.png  -w -nm ranging
/Users/puthier/anaconda3/envs/python_dev_388/lib/python3.8/site-packages/pygtftk-1.5.2-py3.8-macosx-10.9-x86_64.egg/pygtftk/plugins/profile.py:890: FutureWarning: Index.ravel returning ndarray is deprecated; in a future version this will return a view on self.
 |-- 14:40-WARNING-profile : PlotnineError error, can not show group number: Aesthetics {&#39;ha&#39;} specified two times.
</pre></div>
</div>
<a class="reference external image-reference" href="_static/example_08.png"><img alt="_images/example_08.png" src="_images/example_08.png" style="width: 75%;" /></a>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk profile -h
  Usage: gtftk profile -i MATRIX [-o DIR] [-t transcript_file] [-s {mean,median,sum,min,max}] [-e] [-c profile_colors] [-d color_order] [-g {bwig,tx_classes,chrom}] [-f {bwig,tx_classes,chrom}] [-pw page_width] [-ph page_height] [-pf {pdf,png}] [-lw line_width] [-bc border_color] [-x x_lab] [-at axis_text] [-st strip_text] [-u subset_bwig] [-fc facet_col] [-w] [-if user_img_file] [-ul upper_limit] [-nm {none,ranging}] [-tl] [-ti title] [-dpi dpi] [-th {538,bw,grey,gray,linedraw,light,dark,minimal,classic,void,test,matplotlib,seaborn}] [-m palette] [-l] [-h] [-V [verbosity]] [-D] [-C] [-K tmp_dir] [-A] [-L logger_file] [-W write_message_to_file]

  Description: 

     Produces bigWig coverage profiles using calls to plotnine graphic package.

  Notes:
     *  Think about using normalized bigWig files as input to mk_matrix. This will limit the
     requirement for an additional normalization step (see Deeptools for a set of useful methods
     implemented in bamCoverage/bamCompare).

  References:
     *  [1] Numerical Ecology - second Edition - P. Legendre, L. Legendre (1998) Elsevier.

Arguments:
 -i, --inputfile              A zip file containing a matrix as produced by mk_matrix. (default: None)
 -o, --out-dir                Output directory name. (default: draw_profile)
 -t, --transcript-file        A two columns file with the transcripts of interest and their classes. (default: None)
 -s, --stat                   The statistics to be computed. (default: mean)
 -e, --confidence-interval    Add a confidence interval to estimate standard error of the mean. (default: False)
 -c, --profile-colors         Colors. (default: None)
 -d, --color-order            Factor ordering. comma-separated bwig labels or tx classes. (default: None)
 -g, --group-by               The variable used for grouping. (default: None)
 -f, --facet-var              The variable to be used for splitting into facets. (default: None)
 -pw, --page-width            Output pdf file width (e.g. 7 inches). (default: None)
 -ph, --page-height           Output file height (e.g. 5 inches). (default: None)
 -pf, --page-format           Output file format. (default: pdf)
 -lw, --line-width            Line width. (default: 1.25)
 -bc, --border-color          Border color for the plot. (default: #777777)
 -x, --x-lab                  X axis label. (default: Selected genomic regions)
 -at, --axis-text             Size of axis text. (default: 8)
 -st, --strip-text            Size of strip text. (default: 8)
 -u, --subset-bwig            Use only a subset of the bigwigs for plotting (default: None)
 -fc, --facet-col             Number of facet columns. (default: 4)
 -w, --show-group-number      Show the number of element per group (groupe size). (default: False)
 -if, --user-img-file         Provide an alternative path for the image. (default: None)
 -ul, --upper-limit           Upper limit based on quantile computed from unique values. (default: 0.95)
 -nm, --normalization-method  The normalization method performed on a per bigwig basis. (default: none)
 -tl, --to-log                Control whether the data should be log2-transform before plotting. (default: False)
 -ti, --title                 A title for the diagram. (default: )
 -dpi, --dpi                  Dpi to use. (default: 300)
 -th, --theme-plotnine        The theme for plotnine diagram. (default: bw)
 -m, --palette                A color palette (see: https://tinyurl.com/ydacyfxx). (default: nipy_spectral)
 -l, --list-bwig              List the bigwig files in the matrix file.. (default: False)

Command-wise optional arguments:
 -h, --help                   Show this help message and exit.
 -V, --verbosity              Set output verbosity ([0-3]). (default: 0)
 -D, --no-date                Do not add date to output file names. (default: False)
 -C, --add-chr                Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir                Keep all temporary files into this folder. (default: None)
 -A, --keep-all               Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file            Stores the arguments passed to the command into a file. (default: None)
 -W, --write-message-to-file  Store all message into a file. (default: None)
</pre></div>
</div>
</section>
</section>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Commands from section ‘coverage’</a><ul>
<li><a class="reference internal" href="#coverage">coverage</a></li>
<li><a class="reference internal" href="#mk-matrix">mk_matrix</a></li>
<li><a class="reference internal" href="#profile">profile</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="sequence.html"
                        title="previous chapter">Commands from section ‘sequence’</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="miscellaneous.html"
                        title="next chapter">Commands from section ‘miscellaneous’</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/coverage.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="miscellaneous.html" title="Commands from section ‘miscellaneous’"
             >next</a> |</li>
        <li class="right" >
          <a href="sequence.html" title="Commands from section ‘sequence’"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">gtftk 1.5.2 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Commands from section ‘coverage’</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018, F. Lopez and D. Puthier.
      Last updated on Nov 06, 2021.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 4.0.2.
    </div>
  </body>
</html>