
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><script type="text/javascript">

      var _gaq = _gaq || [];
      _gaq.push(['_setAccount', 'UA-139058009-1']);
      _gaq.push(['_trackPageview']);

      (function() {
        var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
        ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
        var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
      })();
    </script>
    <title>Commands from section ‘Editing’ &#8212; gtftk 1.2.1 documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Commands from section ‘selection’" href="selection.html" />
    <link rel="prev" title="Commands from section ‘information’" href="information.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="selection.html" title="Commands from section ‘selection’"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="information.html" title="Commands from section ‘information’"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">gtftk 1.2.1 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Commands from section ‘Editing’</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="commands-from-section-editing">
<h1>Commands from section ‘Editing’<a class="headerlink" href="#commands-from-section-editing" title="Permalink to this headline">¶</a></h1>
<p>In this section we will require the following datasets:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_example -q -d simple -f &#39;*&#39;
</pre></div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_example -q -d mini_real -f &quot;*&quot;
</pre></div>
</div>
<div class="section" id="add-prefix">
<h2>add_prefix<a class="headerlink" href="#add-prefix" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Add a prefix (or suffix) to one of the attribute value (<em>e.g.</em> gene_id)</p>
<p><strong>Example:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk add_prefix -i simple.gtf -k transcript_id -t &quot;novel_&quot;| head -2
chr1	gtftk	gene	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	transcript	125	138	.	+	.	gene_id &quot;G0001&quot;; transcript_id &quot;novel_G0001T002&quot;;
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk add_prefix -h
  Usage: gtftk add_prefix [-i GTF] [-o GTF] [-k KEY] [-t TEXT] [-s] [-f target_feature] [-h] [-V [verbosity]] [-D] [-C] [-K tmp_dir] [-A] [-L logger_file] [-W write_message_to_file]

  Description: 
     *  Add a prefix to target values. By default add &#39;chr&#39; to seqid/chromosome key.

  Version:  2018-01-20

Arguments:
 -i, --inputfile              Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile             Output file. (default: &lt;stdout&gt;)
 -k, --key                    The name of the attribute for which a prefix/suffix is to be added to the corresponding values (e.g, gene_id, transcript_id...). (default: chrom)
 -t, --text                   The character string to add as a prefix to the values. (default: chr)
 -s, --suffix                 The character string to add as a prefix to the values. (default: False)
 -f, --target-feature         The name of the target feature. (default: *)

Command-wise optional arguments:
 -h, --help                   Show this help message and exit.
 -V, --verbosity              Set output verbosity ([0-3]). (default: 0)
 -D, --no-date                Do not add date to output file names. (default: False)
 -C, --add-chr                Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir                Keep all temporary files into this folder. (default: None)
 -A, --keep-all               Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file            Stores the arguments passed to the command into a file. (default: None)
 -W, --write-message-to-file  Store all message into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="del-attr">
<h2>del_attr<a class="headerlink" href="#del-attr" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Delete an attribute and its corresponding values.</p>
<p><strong>Example:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk del_attr -i simple.gtf -k transcript_id,exon_id
chr1	gtftk	gene	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	transcript	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	exon	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	CDS	125	130	.	+	.	gene_id &quot;G0001&quot;; ccds_id &quot;CDS_G0001T002&quot;;
chr1	gtftk	transcript	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	exon	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	CDS	130	132	.	+	.	gene_id &quot;G0001&quot;; ccds_id &quot;CDS_G0001T001&quot;;
chr1	gtftk	gene	180	189	.	+	.	gene_id &quot;G0002&quot;;
chr1	gtftk	transcript	180	189	.	+	.	gene_id &quot;G0002&quot;;
chr1	gtftk	exon	180	189	.	+	.	gene_id &quot;G0002&quot;;
chr1	gtftk	CDS	180	182	.	+	.	gene_id &quot;G0002&quot;; ccds_id &quot;CDS_G0002T001&quot;;
chr1	gtftk	gene	50	61	.	-	.	gene_id &quot;G0003&quot;;
chr1	gtftk	transcript	50	61	.	-	.	gene_id &quot;G0003&quot;;
chr1	gtftk	exon	50	54	.	-	.	gene_id &quot;G0003&quot;;
chr1	gtftk	exon	57	61	.	-	.	gene_id &quot;G0003&quot;;
chr1	gtftk	CDS	50	52	.	-	.	gene_id &quot;G0003&quot;; ccds_id &quot;CDS_G0003T001&quot;;
chr1	gtftk	gene	65	76	.	+	.	gene_id &quot;G0004&quot;;
chr1	gtftk	transcript	65	76	.	+	.	gene_id &quot;G0004&quot;;
chr1	gtftk	exon	65	68	.	+	.	gene_id &quot;G0004&quot;;
chr1	gtftk	exon	71	71	.	+	.	gene_id &quot;G0004&quot;;
chr1	gtftk	exon	74	76	.	+	.	gene_id &quot;G0004&quot;;
chr1	gtftk	CDS	66	68	.	+	.	gene_id &quot;G0004&quot;; ccds_id &quot;CDS_G0004T002&quot;;
chr1	gtftk	CDS	71	71	.	+	.	gene_id &quot;G0004&quot;; ccds_id &quot;CDS_G0004T002&quot;;
chr1	gtftk	CDS	74	75	.	+	.	gene_id &quot;G0004&quot;; ccds_id &quot;CDS_G0004T002&quot;;
chr1	gtftk	transcript	65	76	.	+	.	gene_id &quot;G0004&quot;;
chr1	gtftk	exon	65	68	.	+	.	gene_id &quot;G0004&quot;;
chr1	gtftk	exon	71	71	.	+	.	gene_id &quot;G0004&quot;;
chr1	gtftk	exon	74	76	.	+	.	gene_id &quot;G0004&quot;;
chr1	gtftk	CDS	65	67	.	+	.	gene_id &quot;G0004&quot;; ccds_id &quot;CDS_G0004T001&quot;;
chr1	gtftk	gene	33	47	.	-	.	gene_id &quot;G0005&quot;;
chr1	gtftk	transcript	33	47	.	-	.	gene_id &quot;G0005&quot;;
chr1	gtftk	exon	33	35	.	-	.	gene_id &quot;G0005&quot;;
chr1	gtftk	exon	42	47	.	-	.	gene_id &quot;G0005&quot;;
chr1	gtftk	CDS	43	45	.	-	.	gene_id &quot;G0005&quot;; ccds_id &quot;CDS_G0005T001&quot;;
chr1	gtftk	gene	22	35	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	transcript	22	35	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	exon	22	25	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	exon	28	30	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	exon	33	35	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	CDS	22	25	.	-	.	gene_id &quot;G0006&quot;; ccds_id &quot;CDS_G0006T001&quot;;
chr1	gtftk	CDS	28	30	.	-	.	gene_id &quot;G0006&quot;; ccds_id &quot;CDS_G0006T001&quot;;
chr1	gtftk	CDS	33	34	.	-	.	gene_id &quot;G0006&quot;; ccds_id &quot;CDS_G0006T001&quot;;
chr1	gtftk	transcript	28	35	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	exon	28	30	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	exon	33	35	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	CDS	29	30	.	-	.	gene_id &quot;G0006&quot;; ccds_id &quot;CDS_G0006T002&quot;;
chr1	gtftk	CDS	33	33	.	-	.	gene_id &quot;G0006&quot;; ccds_id &quot;CDS_G0006T002&quot;;
chr1	gtftk	gene	107	116	.	+	.	gene_id &quot;G0007&quot;;
chr1	gtftk	transcript	107	116	.	+	.	gene_id &quot;G0007&quot;;
chr1	gtftk	exon	107	116	.	+	.	gene_id &quot;G0007&quot;;
chr1	gtftk	CDS	112	114	.	+	.	gene_id &quot;G0007&quot;; ccds_id &quot;CDS_G0007T001&quot;;
chr1	gtftk	transcript	107	116	.	+	.	gene_id &quot;G0007&quot;;
chr1	gtftk	exon	107	116	.	+	.	gene_id &quot;G0007&quot;;
chr1	gtftk	CDS	110	115	.	+	.	gene_id &quot;G0007&quot;; ccds_id &quot;CDS_G0007T002&quot;;
chr1	gtftk	gene	210	222	.	-	.	gene_id &quot;G0008&quot;;
chr1	gtftk	transcript	210	222	.	-	.	gene_id &quot;G0008&quot;;
chr1	gtftk	exon	210	214	.	-	.	gene_id &quot;G0008&quot;;
chr1	gtftk	exon	220	222	.	-	.	gene_id &quot;G0008&quot;;
chr1	gtftk	CDS	211	213	.	-	.	gene_id &quot;G0008&quot;; ccds_id &quot;CDS_G0008T001&quot;;
chr1	gtftk	gene	3	14	.	-	.	gene_id &quot;G0009&quot;;
chr1	gtftk	transcript	3	14	.	-	.	gene_id &quot;G0009&quot;;
chr1	gtftk	exon	3	14	.	-	.	gene_id &quot;G0009&quot;;
chr1	gtftk	CDS	5	10	.	-	.	gene_id &quot;G0009&quot;; ccds_id &quot;CDS_G0009T002&quot;;
chr1	gtftk	transcript	3	14	.	-	.	gene_id &quot;G0009&quot;;
chr1	gtftk	exon	3	14	.	-	.	gene_id &quot;G0009&quot;;
chr1	gtftk	CDS	3	8	.	-	.	gene_id &quot;G0009&quot;; ccds_id &quot;CDS_G0009T001&quot;;
chr1	gtftk	gene	176	186	.	+	.	gene_id &quot;G0010&quot;;
chr1	gtftk	transcript	176	186	.	+	.	gene_id &quot;G0010&quot;;
chr1	gtftk	exon	176	186	.	+	.	gene_id &quot;G0010&quot;;
chr1	gtftk	CDS	184	186	.	+	.	gene_id &quot;G0010&quot;; ccds_id &quot;CDS_G0010T001&quot;;
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk del_attr -h
  Usage: gtftk del_attr [-i GTF] [-o GTF] -k KEY [-r] [-v] [-h] [-V [verbosity]] [-D] [-C] [-K tmp_dir] [-A] [-L logger_file] [-W write_message_to_file]

  Description: 
     *  Delete one or several attributes from the gtf file.

  Notes:
     *  You may also use &#39;complex&#39; regexp such as : &quot;(_id)|(_b.*pe)&quot;
     *  Example: gtftk get_example -d mini_real | gtftk del_attr -k &quot;(^.*_id$|^.*_biotype$)&quot; -r -v
     *  TODO: currently a segfault is thrown when no keys are left after deletion (libgtftk issue
     #98).

  Version:  2018-01-20

Arguments:
 -i, --inputfile              Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile             Output file. (default: &lt;stdout&gt;)
 -k, --key                    comma-separated list of attribute names or a regular expression (see -r). (default: None)
 -r, --reg-exp                The key name is a regular expression. (default: False)
 -v, --invert-match           Delected keys are those not matching any of the specified key. (default: False)

Command-wise optional arguments:
 -h, --help                   Show this help message and exit.
 -V, --verbosity              Set output verbosity ([0-3]). (default: 0)
 -D, --no-date                Do not add date to output file names. (default: False)
 -C, --add-chr                Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir                Keep all temporary files into this folder. (default: None)
 -A, --keep-all               Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file            Stores the arguments passed to the command into a file. (default: None)
 -W, --write-message-to-file  Store all message into a file. (default: None)
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In version 1.0.9 gene_id can not be deleted. This will be fixed in next version.</p>
</div>
</div>
<hr class="docutils" />
<div class="section" id="join-attr">
<h2>join_attr<a class="headerlink" href="#join-attr" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Add attributes from a file. This command can be used to import additional key/values into the gtf (e.g CPAT for coding potential, DESeq for differential analysis…). The imported file can be in 2 formats (2 columns or matrix):</p>
<ul class="simple">
<li><p>With a 2-columns file:</p>
<ul>
<li><p>value for joining (transcript_id or gene_id…).</p></li>
<li><p>corresponding value.</p></li>
</ul>
</li>
<li><p>With a matrix (see -m):</p>
<ul>
<li><p>rows corresponding to joining keys (transcript_id or gene_id or…).</p></li>
<li><p>columns corresponding to novel attributes name.</p></li>
<li><p>Each cell of the matrix is a value for the corresponding attribute.</p></li>
</ul>
</li>
</ul>
<p><strong>Example:</strong> With a 2-columns file.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ cat simple.join
G0003	0.2322
G0004	0.999
G0009	0.5555
</pre></div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk join_attr -i simple.gtf -k gene_id -j simple.join -n a_score -t gene| gtftk select_by_key -k feature -v gene
chr1	gtftk	gene	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	gene	180	189	.	+	.	gene_id &quot;G0002&quot;;
chr1	gtftk	gene	50	61	.	-	.	gene_id &quot;G0003&quot;; a_score &quot;0.2322&quot;;
chr1	gtftk	gene	65	76	.	+	.	gene_id &quot;G0004&quot;; a_score &quot;0.999&quot;;
chr1	gtftk	gene	33	47	.	-	.	gene_id &quot;G0005&quot;;
chr1	gtftk	gene	22	35	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	gene	107	116	.	+	.	gene_id &quot;G0007&quot;;
chr1	gtftk	gene	210	222	.	-	.	gene_id &quot;G0008&quot;;
chr1	gtftk	gene	3	14	.	-	.	gene_id &quot;G0009&quot;; a_score &quot;0.5555&quot;;
chr1	gtftk	gene	176	186	.	+	.	gene_id &quot;G0010&quot;;
</pre></div>
</div>
<p><strong>Example:</strong> With a matrix-like file.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ cat simple.join_mat
genes	S1	S2
G0003	0.2322	0.4
G0004	0.999	0.6
G0009	0.5555	0.7
</pre></div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk join_attr -i simple.gtf -k gene_id -j simple.join_mat -m -t gene| gtftk select_by_key -k feature -v gene
chr1	gtftk	gene	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	gene	180	189	.	+	.	gene_id &quot;G0002&quot;;
chr1	gtftk	gene	50	61	.	-	.	gene_id &quot;G0003&quot;; S1 &quot;0.2322&quot;; S2 &quot;0.4&quot;;
chr1	gtftk	gene	65	76	.	+	.	gene_id &quot;G0004&quot;; S1 &quot;0.999&quot;; S2 &quot;0.6&quot;;
chr1	gtftk	gene	33	47	.	-	.	gene_id &quot;G0005&quot;;
chr1	gtftk	gene	22	35	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	gene	107	116	.	+	.	gene_id &quot;G0007&quot;;
chr1	gtftk	gene	210	222	.	-	.	gene_id &quot;G0008&quot;;
chr1	gtftk	gene	3	14	.	-	.	gene_id &quot;G0009&quot;; S1 &quot;0.5555&quot;; S2 &quot;0.7&quot;;
chr1	gtftk	gene	176	186	.	+	.	gene_id &quot;G0010&quot;;
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk join_attr -h
  Usage: gtftk join_attr [-i GTF] [-o GTF] -k KEY -j JOIN_FILE [-H] [-m] [-n NEW_KEY] [-t target_feature] [-h] [-V [verbosity]] [-D] [-C] [-K tmp_dir] [-A] [-L logger_file] [-W write_message_to_file]

  Description: 
     *  Join attributes from a tabulated file.

  Version:  2018-02-05

Arguments:
 -i, --inputfile              Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile             Output file. (default: &lt;stdout&gt;)
 -k, --key-to-join            The name of the key used to join (e.g transcript_id). (default: None)
 -j, --join-file              A two columns file with (i) the value for joining (e.g value for transcript_id), (ii) the value for novel key (e.g the coding potential computed value). (default: None)
 -H, --has-header             Indicates that the &#39;join-file&#39; has a header. (default: False)
 -m, --matrix                 &#39;join-file&#39; expect a matrix with row names as target keys column names as novel key and each cell as value. (default: False)
 -n, --new-key                The name of the novel key. Mutually exclusive with --matrix. (default: None)
 -t, --target-feature         The name(s) of the target feature(s). comma-separated. (default: None)

Command-wise optional arguments:
 -h, --help                   Show this help message and exit.
 -V, --verbosity              Set output verbosity ([0-3]). (default: 0)
 -D, --no-date                Do not add date to output file names. (default: False)
 -C, --add-chr                Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir                Keep all temporary files into this folder. (default: None)
 -A, --keep-all               Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file            Stores the arguments passed to the command into a file. (default: None)
 -W, --write-message-to-file  Store all message into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="join-multi-file">
<h2>join_multi_file<a class="headerlink" href="#join-multi-file" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Join attributes from mutiple files.</p>
<p><strong>Example:</strong> Add key/value to gene features.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ cat simple.join_mat_2
genes	S3	S4
G0003	A	B
G0004	C	D
G0009	E	F
</pre></div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ cat simple.join_mat_3
genes	S5	S6
G0003	0.2322	0.4
G0004	0.999	0.6
G0009	0.5555	0.7
G0009	20	30
G0004	0.999	0.6
</pre></div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk join_multi_file -i simple.gtf -k gene_id -t gene -m simple.join_mat_2 simple.join_mat_3| gtftk select_by_key -g
chr1	gtftk	gene	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	gene	180	189	.	+	.	gene_id &quot;G0002&quot;;
chr1	gtftk	gene	50	61	.	-	.	gene_id &quot;G0003&quot;; S3 &quot;A&quot;; S4 &quot;B&quot;; S5 &quot;0.2322&quot;; S6 &quot;0.4&quot;;
chr1	gtftk	gene	65	76	.	+	.	gene_id &quot;G0004&quot;; S3 &quot;C&quot;; S4 &quot;D&quot;; S5 &quot;0.999|0.999&quot;; S6 &quot;0.6|0.6&quot;;
chr1	gtftk	gene	33	47	.	-	.	gene_id &quot;G0005&quot;;
chr1	gtftk	gene	22	35	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	gene	107	116	.	+	.	gene_id &quot;G0007&quot;;
chr1	gtftk	gene	210	222	.	-	.	gene_id &quot;G0008&quot;;
chr1	gtftk	gene	3	14	.	-	.	gene_id &quot;G0009&quot;; S3 &quot;E&quot;; S4 &quot;F&quot;; S5 &quot;0.5555|20&quot;; S6 &quot;0.7|30&quot;;
chr1	gtftk	gene	176	186	.	+	.	gene_id &quot;G0010&quot;;
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk join_multi_file -h
  Usage: gtftk join_multi_file [-i GTF] [-o GTF] -k KEY [-t target_feature] [-m matrix_files [matrix_files ...]] [-h] [-V [verbosity]] [-D] [-C] [-K tmp_dir] [-A] [-L logger_file] [-W write_message_to_file]

  Description: 
     *  Join attributes from mutiple files.

  Version:  2018-02-05

Arguments:
 -i, --inputfile              Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile             Output file. (default: &lt;stdout&gt;)
 -k, --key-to-join            The name of the key used to join (e.g transcript_id). (default: None)
 -t, --target-feature         The name(s) of the target feature(s). Comma-separated. (default: None)
 -m, --matrix-files           A set of matrix files with row names as target keys column names as novel key and each cell as value. (default: None)

Command-wise optional arguments:
 -h, --help                   Show this help message and exit.
 -V, --verbosity              Set output verbosity ([0-3]). (default: 0)
 -D, --no-date                Do not add date to output file names. (default: False)
 -C, --add-chr                Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir                Keep all temporary files into this folder. (default: None)
 -A, --keep-all               Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file            Stores the arguments passed to the command into a file. (default: None)
 -W, --write-message-to-file  Store all message into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="merge-attr">
<h2>merge_attr<a class="headerlink" href="#merge-attr" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Merge a set of attributes into a destination attribute.</p>
<p><strong>Example:</strong> Merge gene_id and transcript_id into a new key associated to transcript features.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk merge_attr -i simple.gtf -k transcript_id,gene_id -d txgn_id -s &quot;|&quot; -f transcript | gtftk select_by_key -t
chr1	gtftk	transcript	125	138	.	+	.	gene_id &quot;G0001&quot;; transcript_id &quot;G0001T002&quot;; txgn_id &quot;G0001T002|G0001&quot;;
chr1	gtftk	transcript	125	138	.	+	.	gene_id &quot;G0001&quot;; transcript_id &quot;G0001T001&quot;; txgn_id &quot;G0001T001|G0001&quot;;
chr1	gtftk	transcript	180	189	.	+	.	gene_id &quot;G0002&quot;; transcript_id &quot;G0002T001&quot;; txgn_id &quot;G0002T001|G0002&quot;;
chr1	gtftk	transcript	50	61	.	-	.	gene_id &quot;G0003&quot;; transcript_id &quot;G0003T001&quot;; txgn_id &quot;G0003T001|G0003&quot;;
chr1	gtftk	transcript	65	76	.	+	.	gene_id &quot;G0004&quot;; transcript_id &quot;G0004T002&quot;; txgn_id &quot;G0004T002|G0004&quot;;
chr1	gtftk	transcript	65	76	.	+	.	gene_id &quot;G0004&quot;; transcript_id &quot;G0004T001&quot;; txgn_id &quot;G0004T001|G0004&quot;;
chr1	gtftk	transcript	33	47	.	-	.	gene_id &quot;G0005&quot;; transcript_id &quot;G0005T001&quot;; txgn_id &quot;G0005T001|G0005&quot;;
chr1	gtftk	transcript	22	35	.	-	.	gene_id &quot;G0006&quot;; transcript_id &quot;G0006T001&quot;; txgn_id &quot;G0006T001|G0006&quot;;
chr1	gtftk	transcript	28	35	.	-	.	gene_id &quot;G0006&quot;; transcript_id &quot;G0006T002&quot;; txgn_id &quot;G0006T002|G0006&quot;;
chr1	gtftk	transcript	107	116	.	+	.	gene_id &quot;G0007&quot;; transcript_id &quot;G0007T001&quot;; txgn_id &quot;G0007T001|G0007&quot;;
chr1	gtftk	transcript	107	116	.	+	.	gene_id &quot;G0007&quot;; transcript_id &quot;G0007T002&quot;; txgn_id &quot;G0007T002|G0007&quot;;
chr1	gtftk	transcript	210	222	.	-	.	gene_id &quot;G0008&quot;; transcript_id &quot;G0008T001&quot;; txgn_id &quot;G0008T001|G0008&quot;;
chr1	gtftk	transcript	3	14	.	-	.	gene_id &quot;G0009&quot;; transcript_id &quot;G0009T002&quot;; txgn_id &quot;G0009T002|G0009&quot;;
chr1	gtftk	transcript	3	14	.	-	.	gene_id &quot;G0009&quot;; transcript_id &quot;G0009T001&quot;; txgn_id &quot;G0009T001|G0009&quot;;
chr1	gtftk	transcript	176	186	.	+	.	gene_id &quot;G0010&quot;; transcript_id &quot;G0010T001&quot;; txgn_id &quot;G0010T001|G0010&quot;;
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk join_multi_file -h
  Usage: gtftk join_multi_file [-i GTF] [-o GTF] -k KEY [-t target_feature] [-m matrix_files [matrix_files ...]] [-h] [-V [verbosity]] [-D] [-C] [-K tmp_dir] [-A] [-L logger_file] [-W write_message_to_file]

  Description: 
     *  Join attributes from mutiple files.

  Version:  2018-02-05

Arguments:
 -i, --inputfile              Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile             Output file. (default: &lt;stdout&gt;)
 -k, --key-to-join            The name of the key used to join (e.g transcript_id). (default: None)
 -t, --target-feature         The name(s) of the target feature(s). Comma-separated. (default: None)
 -m, --matrix-files           A set of matrix files with row names as target keys column names as novel key and each cell as value. (default: None)

Command-wise optional arguments:
 -h, --help                   Show this help message and exit.
 -V, --verbosity              Set output verbosity ([0-3]). (default: 0)
 -D, --no-date                Do not add date to output file names. (default: False)
 -C, --add-chr                Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir                Keep all temporary files into this folder. (default: None)
 -A, --keep-all               Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file            Stores the arguments passed to the command into a file. (default: None)
 -W, --write-message-to-file  Store all message into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="discretize-key">
<h2>discretize_key<a class="headerlink" href="#discretize-key" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Create a new key by discretizing a numeric key. This can be helpful to create new classes of features on the fly.
The default is to create equally spaced interval. The intervals can also be created by computing the percentiles (-p) which will provide balanced classes most suitable generally.</p>
<p><strong>Example:</strong> Let say we have the following matrix giving expression level of genes (rows) in samples (columns). We could join this information to the GTF and later choose to transform key <em>S1</em> into a new discretized key <em>S1_d</em>. We may apply particular labels to this factor using <em>-l</em>.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ cat simple.join_mat
genes	S1	S2
G0003	0.2322	0.4
G0004	0.999	0.6
G0009	0.5555	0.7
</pre></div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk join_attr -i simple.gtf -j simple.join_mat -k gene_id -m | gtftk discretize_key -k S1 -d S1_d -n 2 -l A,B  | gtftk select_by_key -k feature -v gene
 |-- 22:08-INFO-discretize_key : Categories: [&#39;A&#39;, &#39;B&#39;]
chr1	gtftk	gene	125	138	.	+	.	gene_id &quot;G0001&quot;;
chr1	gtftk	gene	180	189	.	+	.	gene_id &quot;G0002&quot;;
chr1	gtftk	gene	50	61	.	-	.	gene_id &quot;G0003&quot;; S1 &quot;0.2322&quot;; S2 &quot;0.4&quot;; S1_d &quot;A&quot;;
chr1	gtftk	gene	65	76	.	+	.	gene_id &quot;G0004&quot;; S1 &quot;0.999&quot;; S2 &quot;0.6&quot;; S1_d &quot;B&quot;;
chr1	gtftk	gene	33	47	.	-	.	gene_id &quot;G0005&quot;;
chr1	gtftk	gene	22	35	.	-	.	gene_id &quot;G0006&quot;;
chr1	gtftk	gene	107	116	.	+	.	gene_id &quot;G0007&quot;;
chr1	gtftk	gene	210	222	.	-	.	gene_id &quot;G0008&quot;;
chr1	gtftk	gene	3	14	.	-	.	gene_id &quot;G0009&quot;; S1 &quot;0.5555&quot;; S2 &quot;0.7&quot;; S1_d &quot;A&quot;;
chr1	gtftk	gene	176	186	.	+	.	gene_id &quot;G0010&quot;;
</pre></div>
</div>
<p><strong>Example:</strong> We want to load RNA-seq data in the GTF and discretize the expression values according to deciles (-p and -n set to 10). Classes will be labeled from A to J. The example below shows how balanced these classes will be.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p>The <em>profile</em> command that could be used to asses the associated epigenetic marks of these 10 gene classes.</p>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk join_attr -i mini_real.gtf.gz -H -j mini_real_counts_ENCFF630HEX.tsv -k gene_name -n exprs -t gene | gtftk discretize_key -k exprs -p -d exprs_class -n 10 -l A,B,C,D,E,F,G,H,I,J  | gtftk tabulate -k exprs_class -Hn | sort | uniq -c
 |-- 22:08-INFO-discretize_key : Categories: [&#39;A&#39;, &#39;B&#39;, &#39;C&#39;, &#39;D&#39;, &#39;E&#39;, &#39;F&#39;, &#39;G&#39;, &#39;H&#39;, &#39;I&#39;, &#39;J&#39;]
  96 A
  83 B
  89 C
  91 D
  88 E
  88 F
  90 G
  89 H
  89 I
  90 J
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk discretize_key -h
  Usage: gtftk discretize_key [-i GTF] [-o GTF] -k src_key -d dest_key -n KEY [-l labels] [-p] [-g] [-u] [-r precision] [-h] [-V [verbosity]] [-D] [-C] [-K tmp_dir] [-A] [-L logger_file] [-W write_message_to_file]

  Description: 
     *  Create a new key by discretizing a numeric key. This can be helpful to create new classes on the
     fly that can be used subsequently.

  Notes:
     *  if --ft-type is not set the destination key will be assigned to all feature containing the
     source key.
     *  Non-numeric value for source key will be translated into &#39;NA&#39;.
     *  The default is to create equally spaced interval. The interval can also be created by
     computing the percentiles (-p).

  Version:  2018-01-20

Arguments:
 -i, --inputfile              Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile             Output file. (default: &lt;stdout&gt;)
 -k, --src-key                The name of the source key (default: None)
 -d, --dest-key               The name of the target key. (default: None)
 -n, --nb-levels              The number of levels/classes to create. (default: 2)
 -l, --labels                 A comma-separated list of labels of size --nb-levels. (default: None)
 -p, --percentiles            Compute --nb-levels classes using percentiles. (default: False)
 -g, --log                    Compute breaks based on log-scale. (default: False)
 -u, --percentiles-of-uniq    Compute percentiles based on non-redondant values. (default: False)
 -r, --precision              The precision used in naming intervals. (default: 2)

Command-wise optional arguments:
 -h, --help                   Show this help message and exit.
 -V, --verbosity              Set output verbosity ([0-3]). (default: 0)
 -D, --no-date                Do not add date to output file names. (default: False)
 -C, --add-chr                Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir                Keep all temporary files into this folder. (default: None)
 -A, --keep-all               Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file            Stores the arguments passed to the command into a file. (default: None)
 -W, --write-message-to-file  Store all message into a file. (default: None)
</pre></div>
</div>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Commands from section ‘Editing’</a><ul>
<li><a class="reference internal" href="#add-prefix">add_prefix</a></li>
<li><a class="reference internal" href="#del-attr">del_attr</a></li>
<li><a class="reference internal" href="#join-attr">join_attr</a></li>
<li><a class="reference internal" href="#join-multi-file">join_multi_file</a></li>
<li><a class="reference internal" href="#merge-attr">merge_attr</a></li>
<li><a class="reference internal" href="#discretize-key">discretize_key</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="information.html"
                        title="previous chapter">Commands from section ‘information’</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="selection.html"
                        title="next chapter">Commands from section ‘selection’</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/editing.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="selection.html" title="Commands from section ‘selection’"
             >next</a> |</li>
        <li class="right" >
          <a href="information.html" title="Commands from section ‘information’"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">gtftk 1.2.1 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">Commands from section ‘Editing’</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018, F. Lopez and D. Puthier.
      Last updated on Sep 16, 2020.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.2.1.
    </div>
  </body>
</html>