
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Commands from section ‘sequence’ &#8212; gtftk 1.0.5 documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Commands from section ‘coverage’" href="coverage.html" />
    <link rel="prev" title="Commands from section ‘coordinates’" href="coordinates.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="coverage.html" title="Commands from section ‘coverage’"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="coordinates.html" title="Commands from section ‘coordinates’"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">gtftk 1.0.5 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="commands-from-section-sequence">
<h1>Commands from section ‘sequence’<a class="headerlink" href="#commands-from-section-sequence" title="Permalink to this headline">¶</a></h1>
<p>In this section we will require the following datasets:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_example -q -d simple -f &#39;*&#39;
</pre></div>
</div>
<div class="section" id="get-tx-seq">
<h2>get_tx_seq<a class="headerlink" href="#get-tx-seq" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Get transcript sequences in fasta format.</p>
<p><strong>Example:</strong> Get sequences of transcripts in 5’ to 3’ orientation</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_tx_seq -g simple.fa -i simple.gtf | head -n 4
&gt;transcript|G0001T002|G0001|chr1|125|138
cccccgttacgtag
&gt;transcript|G0001T001|G0001|chr1|125|138
cccccgttacgtag
</pre></div>
</div>
<p>Note that the format is rather flexible and any combination of key can be exported to the header.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_tx_seq -i simple.gtf -g simple.fa  -l gene_id,transcript_id,feature,chrom,start,end,strand  | head -n 2
&gt;G0001|G0001T002|transcript|chr1|125|138|+
cccccgttacgtag
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_tx_seq -h
  Usage: gtftk get_tx_seq [-i GTF] [-o FASTA] -g FASTA [-w] [-s SEP] [-l label] [-f] [-d] [-a assembly] [-c] [-n] [-e] [-h] [-V ] [-D] [-C] [-K] [-A] [-L]

  Description: 
     *  Get transcripts sequences in a flexible fasta format from a GTF file.

  Notes:
     *  The sequences are returned in 5&#39; to 3&#39; orientation.
     *  If you want to use wildcards, use quotes :e.g. &#39;foo/bar*.fa&#39;.
     *  The first time a genome is used, an index (*.fa.gtftk) will be created in ~/.gtftk.

  Version:

Arguments:
 -i, --inputfile       Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile      Output FASTA file. (default: &lt;stdout&gt;)
 -g, --genome          The genome in fasta format. Accept path with wildcards (e.g. *.fa). (default: None)
 -w, --with-introns    Set to true to include intronic regions. (default: False)
 -s, --separator       To separate info in header. (default: |)
 -l, --label           A set of key for the header. (default: feature,transcript_id,gene_id,seqid,start,end)
 -f, --sleuth-format   Produce output in sleuth format (still experimental). (default: False)
 -d, --delete-version  In case of --sleuth-format, delete gene_id or transcript_id version number (e.g &#39;.2&#39; in ENSG56765.2). (default: False)
 -a, --assembly        In case of --sleuth-format, an assembly version. (default: GRCm38)
 -c, --del-chr         When using --sleuth-format delete &#39;chr&#39; in sequence id. (default: False)
 -n, --no-rev-comp     Don&#39;t reverse complement sequence corresponding to gene on minus strand. (default: False)
 -e, --explicit        Write explicitly the name of the keys in the header. (default: False)

Command-wise optional arguments:
 -h, --help            Show this help message and exit.
 -V, --verbosity       Set output verbosity ([0-3]). (default: 0)
 -D, --no-date         Do not add date to output file names. (default: False)
 -C, --add-chr         Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir         Keep all temporary files into this folder. (default: None)
 -A, --keep-all        Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file     Stores the arguments passed to the command into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="get-feat-seq">
<h2>get_feat_seq<a class="headerlink" href="#get-feat-seq" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Get feature sequence (e.g exon, UTR…).</p>
<p><strong>Example:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_feat_seq -i simple.gtf -g simple.fa  -l feature,transcript_id,start -t  exon -n | head -10
&gt;exon|G0001T002|124
cccccgttacgtag
&gt;exon|G0001T001|124
cccccgttacgtag
&gt;exon|G0002T001|179
ggccttatta
&gt;exon|G0003T001|49
caagc
&gt;exon|G0003T001|56
taatt
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_feat_seq -h
  Usage: gtftk get_feat_seq [-i GTF] [-o FASTA] -g FASTA [-s separator] [-l label] [-t feature_type] [-n] [-r] [-u] [-h] [-V ] [-D] [-C] [-K] [-A] [-L]

  Description: 
     *  Get feature sequences (i.e. column 3) in a flexible fasta format from a GTF file.

  Notes:
     *  The sequences are returned in 5&#39; to 3&#39; orientation.
     *  If you want to use wildcards, use quotes: e.g. &#39;foo/bar*.fa&#39;.
     *  See get_tx_seq for mature RNA sequence.
     *  If --unique is used if a header was already encountered the record won&#39;t be print.  Take
     care to use unambiguous identifiers in the header.

  Version:

Arguments:
 -i, --inputfile           Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile          Output FASTA file. (default: &lt;stdout&gt;)
 -g, --genome              The genome in fasta format. (default: None)
 -s, --separator           To separate info in header. (default: |)
 -l, --label               A set of key for the header that will be extracted from the transcript line. (default: feature,transcript_id,gene_id,seqid,start,end)
 -t, --feature-type        The feature type (one defined in column 3). (default: exon)
 -n, --no-rev-comp         Don&#39;t reverse complement sequence corresponding to gene on minus strand. (default: False)
 -r, --rev-comp-to-header  Indicate in the header whether sequence was rev-complemented. (default: False)
 -u, --unique              Don&#39;t write redondant IDS. (default: False)

Command-wise optional arguments:
 -h, --help                Show this help message and exit.
 -V, --verbosity           Set output verbosity ([0-3]). (default: 0)
 -D, --no-date             Do not add date to output file names. (default: False)
 -C, --add-chr             Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir             Keep all temporary files into this folder. (default: None)
 -A, --keep-all            Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file         Stores the arguments passed to the command into a file. (default: None)
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Commands from section ‘sequence’</a><ul>
<li><a class="reference internal" href="#get-tx-seq">get_tx_seq</a></li>
<li><a class="reference internal" href="#get-feat-seq">get_feat_seq</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="coordinates.html"
                        title="previous chapter">Commands from section ‘coordinates’</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="coverage.html"
                        title="next chapter">Commands from section ‘coverage’</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/sequence.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="coverage.html" title="Commands from section ‘coverage’"
             >next</a> |</li>
        <li class="right" >
          <a href="coordinates.html" title="Commands from section ‘coordinates’"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">gtftk 1.0.5 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018, F. Lopez and D. Puthier.
      Last updated on Apr 02, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.3.
    </div>
  </body>
</html>